[tituschewxj] iP

tituschewxj opened the pull request:
# Stratify: Elevate Your Task Management :clipboard:

>_Stratify is more than just a task organizer; it's a seamless experience designed to amplify your productivity._ 

Here's why it stands out:

* :computer: **Command-Line Simplicity**: A user-friendly, command-line-input based interface ensures efficient task handling without unnecessary complexities.

* :watch: **Effective Task Management**: Stratify empowers you to effortlessly manage your tasks, providing a streamlined approach to boost your productivity. 

* :bust_in_silhouette: **User-Friendly Design**: Easy to use, Stratify eliminates the learning curve. Dive in, follow a few simple steps, and let it transform how you manage your tasks. 

## Get Started in Four Steps:
1. **Clone the Repository**: Fetch Stratify to your local machine effortlessly from [here](https://github.com/tituschewxj/ip).
1. **Compile and Run**: Quick setup with minimal effort.
1. **Add Your Tasks**: Seamlessly integrate your to-dos, deadlines, and events with just a few commands.
1. **Let it Manage for You**: Sit back as Stratify efficiently orchestrates your tasks, allowing you to focus on what truly matters.

## Features That Set Stratify Apart:
- [x] :memo: **Task Management**: Efficiently organize your to-dos.
- [x] :date: **Deadline Tracking**: Stay ahead by managing crucial deadlines.
- [x] :people_holding_hands: **Event Coordination**: Seamlessly integrate and manage your events.
- [x] :leftwards_arrow_with_hook: **Made a mistake?** Undo your changes! 


## Calling All Java Enthusiasts:
Are you a Java programmer looking for a practical playground? Stratify has you covered. Enhance your Java skills by running it's simple `main` method:
```java
public static void main(String[] args) {
    Application.launch(ChatBot.class, args);
}
```

>Elevate your task management with Stratify — where simplicity meets productivity.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tituschewxj committed:
Add ChatBot.java

Add a class to encapsulate the behaviour of a Chat Bot.

tituschewxj committed:
Update Duke main method

tituschewxj committed:
ChatBot.java: Add indentation and read input

tituschewxj committed:
Duke.java: Run chatbot

tituschewxj committed:
ChatBot.java: Add adding and listing functionality

Parsing of input with commands Enum is used to help identify which command is being called.

tituschewxj committed:
ChatBot.java: Refactor indentation code

tituschewxj committed:
Task.java: Create Task Class

tituschewxj committed:
ChatBot.java: Parse mark and unmark commands

tituschewxj committed:
Create TaskList Class

tituschewxj committed:
Refactor if-else conditions

tituschewxj committed:
Add Deadline Class

tituschewxj committed:
Add ToDo Class

tituschewxj committed:
Add Event Class

tituschewxj committed:
Deadline.java: Fix toString

tituschewxj committed:
Add Argument Class

This class encapsulates an argument and it's value.

tituschewxj committed:
Refactor Command Enum to separate file

tituschewxj committed:
Add Action class

This class encapsulates the command and it's arguments.

tituschewxj committed:
Add InputParser class

This class handles the parsing of the input and it's arguments.

tituschewxj committed:
TaskList.java: Add todo, deadline, event tasks

tituschewxj committed:
TaskList.java: Add methods

Add methods to get most recent task and get number of tasks.

tituschewxj committed:
Chatbot.java: Refactor away Command Enum

tituschewxj committed:
Chatbot.java: Refactor away input parsing

tituschewxj committed:
Chatbot.java: Add Actions to executable operations

Abstract away the complexity of the commands and arguments with Action objects.

tituschewxj committed:
InputParser.java: Trim whitespace in arguments

tituschewxj committed:
Argument.java: Method to match argument names

tituschewxj committed:
Add Printer class

Add Printer.java to encapsulate the printing of ChatBot messages to the console.

tituschewxj committed:
Add Automated Text UI Testing

tituschewxj committed:
Handle invalid command

tituschewxj committed:
Rename methods and enums

To make the ChatBot methods and enums names more consistent and readable.

tituschewxj committed:
Remove add command

Adding a generic task no longer happens due to the implementation of invalid commands.

tituschewxj committed:
Update Commands enum with name and usage

Store command names and usage in the enum.

tituschewxj committed:
Add validation to commands

tituschewxj committed:
Refactor Command to use Arguments

Command isn't able to validate an Action, but requires a separate check in ChatBot.java.

Using Arguments allows for implementing more versatile commands, and simplifies the process of validating of the arguments.

Let's,
* Command take in Arguments that define its name and usage.

tituschewxj committed:
Update test cases

tituschewxj committed:
Refactor and restructure codebase

Making subtypes of Action that perform behaviour related to each command follows the principles object-oriented programming.

Let's,
* Make subclasses for Action to encapsulate the behaviour of different actions.
* Move classes into their packages to organize the codebase.

tituschewxj committed:
Command.java: Update generateUsageHint method

Let's remove the spacing at the end of the usage hint.

tituschewxj committed:
Update javadocs

tituschewxj committed:
Add Exceptions to handle unexpected input

tituschewxj committed:
Use exceptions to handle errors

InvalidAction.java is removed as an exception handles the state when an invalid argument is entered.

Let's,
* Remove InvalidAction.java
* Use Exceptions to handle errors

tituschewxj committed:
Add delete task command

tituschewxj committed:
action: Refactor handleAddSuccess

tituschewxj committed:
action.exception: Update exception messages

tituschewxj committed:
text-ui-test: Update test cases

tituschewxj committed:
Refactor: Move Argument and Command to action.util

tituschewxj committed:
InputParser.java: Improve input parsing

Only when a word starts with "/", will it be considered as the start of a new argument.

tituschewxj committed:
UnrecognizedArgumentException.java: Refactor constructor arguments

tituschewxj committed:
Move validation of missing values to Action constructor

This refactoring makes the code cleaner.

Let's,
* Move validation of missing arguments to Action constructor
* Make variables in Argument and Command private
* Add UnexpectedArgumentValueException.java
* Update EXPECTED.TXT with new behaviour from UnexpectedArgumentValueException

tituschewxj committed:
action package: Remove throw exception when no exception is thrown

tituschewxj committed:
Action.java: Add factory method

The logic for determining the type of Action created should not be in InputParser.java, but in Action.java.

tituschewxj committed:
Add ExpectedArgument.java

Since Arguments in Command.java are being misused as expected arguments rather than supplied arguments.

tituschewxj committed:
Validate in Command.java

Move the logic of validateExpectedArguments to Command class.

Command should encapsulate the logic of validating expected arguments.

Validation of arguments is now in one less place.

tituschewxj committed:
Add save feature

Saves the tasks in the hard disk automatically whenever the task list changes. The tasks are saved as a human-readable list, in order for the user to read their list using the save file.

Changes are made to support the new save feature.

Let's,
* save the task list automatically
* update tasks to support reading of a task string
* add LocalStorage.java to encapsulate storing of files
* add a print queue to store messages such as the message when a save file is found, which can be printed together with the chat message
* add ./data/ folder to git ignore

tituschewxj committed:
Merge save with master

tituschewxj committed:
Store dates as java.time.LocalDate

tituschewxj committed:
Merge branch 'branch-Level-8'

tituschewxj committed:
Organize into packages

tituschewxj committed:
Merge branch 'branch-A-Packages'

tituschewxj committed:
Add StringValue.java

Add StringValue which is a supertype of DateStringValue.java

tituschewxj committed:
Remove Command Enum

Remove Command enums as it is not useful to have such enums, puts an Action's logic in two different files.

Let's replace with a constant COMMAND in each Action subclass with the expected arguments.

tituschewxj committed:
Set classes as final

tituschewxj committed:
Add more exceptions

Move the functionality of validating to the TaskList/IntegerStringValue.

Let's,
* add ChatBotException
* add OutOfBoundsException
* add InvalidValueTypeException
* add InvalidTaskStringException

tituschewxj committed:
Add TaskParser.java

Move parsing of tasks from chatbot.task to chatbot.parse

tituschewxj committed:
Merge branch 'branch-A-MoreOOP'

tituschewxj committed:
Merge branch add-gradle-support

tituschewxj committed:
Add unit tests

Use JUnit to add tests.

tituschewxj committed:
Merge branch 'branch-A-JUnit'

tituschewxj committed:
Update Javadocs

tituschewxj committed:
Update code standard

tituschewxj committed:
Add find command

Allows users to find a task by searching using a regex pattern

tituschewxj committed:
Merge branch 'branch-A-JavaDoc'

tituschewxj committed:
Merge branch 'branch-A-CodingStandard'

tituschewxj committed:
Merge branch 'branch-Level-9'

tituschewxj committed:
Add runShadow to build.gradle

tituschewxj committed:
Refactor IntegerStringValue.java

tituschewxj committed:
Refactor DateStringValue.java

tituschewxj committed:
Add checkstyle configuration

tituschewxj committed:
Fix checkstyle issues

tituschewxj committed:
Merge branch 'branch-A-CheckStyle'

tituschewxj committed:
FindAction.java: Rename variable

tituschewxj committed:
Add JavaFX Part 1

tituschewxj committed:
Add JavaFX Part 2

tituschewxj committed:
Add JavaFX Part 3

tituschewxj committed:
Refactor Printer.java to PrintFormatter.java

tengcharmaine commented on src/main/java/chatbot/ChatBot.java:
```diff
@@ -0,0 +1,70 @@
+        userList = LocalStorage.loadTaskList();
+    }
+
+    public static void main(String[] args) {
```
Can add JavaDocs to this method? I am not too sure about this since the guide only mentioned about overriding and getters/setters being exceptions to this.

tengcharmaine commented on src/main/java/chatbot/parse/InputParser.java:
```diff
@@ -0,0 +1,85 @@
+import chatbot.action.util.Argument;
+import chatbot.action.util.Command;
+
+
+
```
Can remove these 2 lines! But is something very minor.

tengcharmaine reviewed (commented):
LGTM! Just some minor changes suggested.

tituschewxj committed:
Style GUI using JavaFX

tituschewxj committed:
Merge branch 'branch-Level-10'

tituschewxj committed:
Add assertions for assumed non-null values

Assertions are used to define assumptions about the program state so that the runtime can verify them.

Let's
* add assertions to arguments that cannot be null.

tituschewxj committed:
Improve code quality

Improve the code quality following Code Quality topics.

Code quality is important to improving the readability of the code.

Let's
* Use guard clauses where necessary
* Update JavaDocs to follow coding standard guidelines
* Extract out method for actions that use an index
* Use named constants when suitable

tituschewxj committed:
Use Streams

Let's use the Streams feature of Java, where applicable

tituschewxj committed:
Merge pull request #2 from tituschewxj/branch-A-CodeQuality

Improve code quality

tituschewxj committed:
Merge pull request #1 from tituschewxj/branch-A-Assertions

Add assertions

tituschewxj committed:
Resolve merge conflict

tituschewxj committed:
Merge pull request #3 from tituschewxj/branch-A-Streams

Use Streams

tituschewxj committed:
Set up CI

Use GitHub Actions to set up Continuous Integration (CI).

tituschewxj committed:
Merge pull request #4 from tituschewxj/branch-A-CI

Set up CI

SamuelZero1 commented on src/main/java/TaskList.java:
public void markTask(int index) {
        if (isValidIndex(index)) tasks.get(index).mark();
    }
    
should change to:
public void markTask(int index) {
        if (isValidIndex(index)) {
            tasks.get(index).mark();
        }
...

tituschewxj committed:
Add Undo Command

Let's
* add undo functionality
* add a store to save previous states of the task list

tituschewxj committed:
Improve Undo Command

Let's
* display the change that is undone

tituschewxj committed:
Undo by rolling back by a number of steps

Undo takes in an OptionalArgument that specifies the number of steps to undo by.

Also improve the design of Arguments by having optional and supplied arguments.

Let's
* add OptionalArgument to handle arguments that can be expected, by are not required
* add SuppliedArgument to separate the behaviour from Argument
* update Undo to undo by a number of steps
* move validation logic for argument value types to SuppliedArgument

tituschewxj committed:
Merge pull request #5 from tituschewxj/branch-C-Undo

Add Undo Command

tituschewxj committed:
Add resize functionality to UI

tituschewxj committed:
Add color from message type

tituschewxj committed:
Merge pull request #7 from tituschewxj/branch-A-BetterGui

Add better GUI

tituschewxj committed:
Add User Guide

tituschewxj committed:
Merge pull request #8 from tituschewxj/branch-A-UserGuide

Add User Guide

tituschewxj committed:
Fix docs formatting

tituschewxj committed:
Fix docs formatting

tituschewxj committed:
Fix docs formatting

tituschewxj committed:
Fix docs formatting

tituschewxj committed:
Fix docs formatting

tituschewxj committed:
LocalStorage: Init save when no previous save

Let's save the state when no save file is found at the start

tituschewxj committed:
Update GitHub pages README.md
